src/stream/src/executor/managed_state/join/join_entry_state.rs

```diff
@@ -52,19 +53,43 @@ impl JoinEntryState {
         key: PkType,
         value: StateValueType,
     ) -> Result<&mut StateValueType, JoinEntryError> {
+        let mut removed = false;
+        if !enable_strict_consistency() {
+            // strict consistency is off, let's remove existing (if any) first
+            if let Some(old_value) = self.cached.remove(&key) {
+                self.kv_heap_size.sub(&key, &old_value);
+                removed = true;
+            }
+        }
+
         self.kv_heap_size.add(&key, &value);
-        self.cached
-            .try_insert(key, value)
-            .map_err(|_| JoinEntryError::OccupiedError)
+
+        let ret = self.cached.try_insert(key.clone(), value);
+
+        if !enable_strict_consistency() {
+            assert!(ret.is_ok(), "we have removed existing entry, if any");
+            if removed {
+                // if not silent, we should log the error
+                tracing::error!(?key, "double inserting a join state entry");
+            }
+        }
+
+        ret.map_err(|_| JoinEntryError::OccupiedError)
     }
 
     /// Delete from the cache.
     pub fn remove(&mut self, pk: PkType) -> Result<(), JoinEntryError> {
         if let Some(value) = self.cached.remove(&pk) {
             self.kv_heap_size.sub(&pk, &value);
             Ok(())
-        } else {
+        } else if enable_strict_consistency() {
             Err(JoinEntryError::RemoveError)
+        } else {
+            tracing::error!(
+                ?pk,
+                "removing a join state entry but it is not in the cache"
+            );
+            Ok(())
```

Comment:
I'm considering whether it's possible to further reduce such ad-hoc path for handling errors in non-strict mode, with a similar idea to `inconsistency_panic!`. That is to say, can we do this at the **callsite**:

```rust
entry
    .remove(pk)
    .with_context(|| self.state.error_context(&value))
    .maybe_tolerate_inconsistency("removing a join state entry but it is not in the cache")
?;
```

and `maybe_tolerate_inconsistency` (random name) is like

```rust
impl<E> Result<(), E> {
    fn maybe_tolerate_inconsistency<T, E>(self, msg: &str) -> Self {
        if !enable_strict_consistency() && let Err(e) = self {
            tracing::error!(error = %e.as_report(), "{}", msg);
            Ok(())
        } else {
            self
        }
    }
}
```

However, it seems to only work with `Result` of unit type `Ok`. 😄